### ISSUE TITLE
Generated C++ parser code does not compile for 32-bits build

### ISSUE THREAD

**Issue by reiniert** [GaloisInc/daedalus#159]
When trying to compile the generated parser code as 32 bits (with the flag `-m32`), the following assertion fails:
```
daedalus/rts-c/ddl/cast.h:44:3: error: static_assert failed due to requirement 'sizeof(unsigned long) * 8 >= 64' "Unsupported cast: uint_to_integer"
   static_assert(sizeof(unsigned long) * 8 >= in,
   ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
The 32 bits build is desired because address sanitation is enabled and a 64 bits build in that case will allocate too much memory.

**Comment by yav** [comment]
Good point, I was kind of assume a 64-bit architecture, I'll have to go through the RTS to see if I can make it work for 32 bits.   I don't have a lot of time to work on it right now though, is this a blocker, or more of a "nice to have"?

**Comment by reiniert** [comment]
Without Address Sanitation, the memory errors that I am supposed to find will not be detectable. I have not been able to run this with Address Sanitation in 64-bits mode, due to the large amount of memory it requires. So for me, this is a blocker indeed, sorry about that :-)

**Comment by yav** [comment]
OK, I'll have a look in the next few days.  I think the main challenge I can think if is figure how to interface with `libgmp`

**Comment by reiniert** [comment]
Thanks. In the past I was able to build generated C++ parser code with `-m32` and at that time I already ran into a dependency on `libgmp`. It did not seem to be a problem: using Ubuntu, I just had to install the package `libgmp-dev:i386`. I never saw that `static_assert` fail though (but I was using a different DDL).

**Comment by yav** [comment]
Yeah, I think it should build just fine.  The main issue I need to work around is how to convert from a `libgmp` big integer to a `uint 64`.  `libgmp` has a function to give you an `unsigned long`, and I was assuming that this would be 64 bits, but I can add some more smarts to do something different depending on `sizeof(unsigned long)`.

**Comment by yav** [comment]
@reiniert while I am working on 32-bit support, does using Valgrind help with looking for memory errors, or does it also fall over on the examples you you using?

**Comment by reiniert** [comment]
I have not tried valgrind, I do not think it plays well with the fuzzer (AFL++) which I use to find the problems. It may be good for triaging the crashes once found though.

**Comment by yav** [comment]
@reiniert sorry for the delay, I just came back from vacation.  This is certainly on my TODO list, and I am hoping to have something working fairly soon, and I'll update the ticket when I do.

**Comment by reiniert** [comment]
@yav I finally got to try this and my code seems to compile with `-m32` now, thank you. I wrote "seems to" because I have not actually seen the build step complete because my machine is running out of memory. This may be due to the generated C++ parser code being a little under 30K lines long and the fuzzer version of the compiler is ram-greedy. I will explore further.
